### PR TITLE
feat(contracts): include DSTs when hashing leaves

### PR DESCRIPTION
**Pull request by arajasek**
Adds one-byte DSTs to our hashing, so far using 1 for XBlockHeaders and 2 for XMsgs.

Note that this does not implement full EIP-712, because it's hard to shoehorn what we need into that format (eg. what's the Chain ID to use?).

issue: #1433

### CODE REVIEW

File: lib/xchain/merkle.go [hunk 1]
```diff
@@ -7,6 +7,11 @@ import (
 	"github.com/ethereum/go-ethereum/common"
 )
 
+const (
+	DomainSeparationTagBlockHeader merkle.DomainSeparationTag = 1 + iota
+	DomainSeparationTagMessage
```
Comment (corverroos) [inline lib/xchain/merkle.go:12]: lets explicitly number these enums. I suggest to only use `iota` for in-memory enums. When persisting enums to disk, it is better to be super explicit about there values, this avoid accidental changing of values due reordering or deprecation. Also improves lookup. Don't need to count in your head.

TL;DR: only use iota if the actual number doesn't matter.

Comment (corverroos) [reply lib/xchain/merkle.go:12]: Can maybe make names more concise, especially since the type is already full qualified.
```
const (
    DSTUnknown  merkle.DomainSeparationTag = 0
    DSTBlockHeader merkle.DomainSeparationTag = 1
    DSTMessage merkle.DomainSeparationTag = 2
)
```

**Review by corverroos — APPROVED** [review]

File: contracts/core/src/libraries/XBlockMerkleProof.sol [hunk 2]
```diff
@@ -9,6 +9,10 @@ import { XTypes } from "./XTypes.sol";
  * @dev Library for verifying XBlock merkle proofs
  */
 library XBlockMerkleProof {
+    uint8 internal constant DST_XBLOCK_HEADER = 1;
+
+    uint8 internal constant DST_XMSG = 2;
+
```
Comment (kevinhalliday) [inline contracts/core/src/libraries/XBlockMerkleProof.sol:15]: Let's add some natspec comments here.

Also thinking - these aren't exaclty "domain separators" (is DST == domain separator? or domain separator type?)  They are kinda like the type hashes used in EIP-712 domain separators. But they also aren't type hashes. 

Is there another name we can give these? like `_TYPE_ID`? or maybe `_HASH_PREFIX`?

**Review by kevinhalliday — COMMENTED** [review]
Just need to update the ts test data generator

File: contracts/core/src/libraries/XBlockMerkleProof.sol (hunk 2 above)
Comment (arajasek) [reply contracts/core/src/libraries/XBlockMerkleProof.sol]: Comments added.

I think these are domain separators (DST = domain separation tag), unless I'm very mistaken.

Comment (kevinhalliday) [reply contracts/core/src/libraries/XBlockMerkleProof.sol]: I guess my understanding of the semantics of "domain separators" was that they were more about verifying chain & contract. but types are a domain in themselves. so this works

**Review by kevinhalliday — APPROVED** [review]
